### DISCUSSION TITLE
troubleshooting/faq

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#27] [General]
# troubleshooting/faq

Commonly asked questions about Velopack.

https://docs.velopack.io/troubleshooting/faq

**Comment by ConnectAutomatisering** [comment]
Hi,

I have the following challenge.
My update url is dynamic. So the application is running for multiple customers that have different update url's because its running locally.
What would be a good solution for that?

The created exe is hosted via a .net service, so the url is known there.

**Reply by caesay** [reply]
You are in control, as you pass in the desired update url to update manager via `new UpdateManager(...source...)`. So just change what you pass in to the constructor?

**Reply by ConnectAutomatisering** [reply]
That i understand. The thing is what would be a good way to facilitate that?

For example: Before the .exe is downloaded from the api the URL/(some metadata) can be added to the .exe so the url is embedded in the installer after the installer is created. This is the way we got it to work first with Clickonce. The manifest that is generated on the fly so we could pass a dynamic url.

I was thinking about adding metadata to the installer file orso but i havent figured out a good solution and i would like to prevent to build a velopack version per customer.

**Reply by caesay** [reply]
It is not supported to modify a Velopack installer in any way after it's been created, as this would break the code signature or worse - corrupt the bundled application. You could ask your users for a customer-specific key on startup which then downloads your manifest.

**Reply by ConnectAutomatisering** [reply]
Thank you for you're responses. Very valid arguments. Im gonna think about it. We would like the installation to be as easy as possible. If i find a solution ill put in the comments.

**Comment by Ayazis** [comment]
Hi,

I've looked in the documentation and the available hooks / options. I haven't find the answer to these two questions:
- Is there an option to present the user with a EULA before continuing installation? 
- Is there an option to present the user with an custom installation folder?

**Reply by caesay** [reply]
Both of these are possible on macos, but not yet on windows. See https://github.com/velopack/velopack/issues/30 where this is discussed / being worked on for Windows.

**Comment by Ananym** [comment]
Was never able to use squirrel very well because constantly changing the actual executable path plays havoc with firewall and security exceptions etc etc.  Does Velopack have a solution for that?

**Reply by caesay** [reply]
Velopack doesn't do that. Your application will stay at the same location through updates.

**Reply by Ananym** [reply]
Awesome - thanks!

**Comment by kalvinet** [comment]
Is it possible to configure a preinstall script on macOS to ensure .NET is installed before the application?

**Reply by caesay** [reply]
We do support [bootstrapping dotnet on windows](https://docs.velopack.io/packaging/bootstrapping), however, it's currently not possible to bootstrap dotnet on mac, nor is it possible to provide a custom preinstall script. 

Our current recommendation on linux and macos is to ship dotnet self contained, as framework-dependent installs tend to have far too many compatibility/runtime issues. I am not interested in adding support for it at this time either. The ecosystem/process for installing dotnet on linux/mac is just very messy, inconsistent, and regularly requires manual fixing on the command line before it's actually usable.

If you absolutely require this feature, I'd suggest you take/extract our portable zip output and create your own dmg/pkg with whatever custom scripts you need.

**Reply by kalvinet** [reply]
Thanks for the recommendations. My obfuscator has issues with self contained files, I will try to create a custom dmg/pkg.

**Comment by Levvy055** [comment]
How to add URI Protocols in Windows registry? 
Want to have a hook to start app from browser URI like myapp://settings

**Reply by caesay** [reply]
We eventually plan to support it via https://github.com/velopack/velopack/issues/29. Until that issue is complete, there is no built in support, so you'd need to DIY and install the registry keys yourself using the [VelopackApp](https://docs.velopack.io/reference/cs/Velopack/VelopackApp) hooks.

**Comment by metscore** [comment]
Hi,

Forgive me if I missed this in the documentation, but is it possible to select to install for *all users* instead of *per user*?
I have a lab machine that multiple users are using (with their own Windows accounts) and it would be very convenient if the application could be installed for all, instead of having 10-ish parallel installations on the same Windows 11 machine (which the fact of our current ClickOnce installation)

**Reply by caesay** [reply]
This will be supported once https://github.com/velopack/velopack/pull/606 is merged.

**Comment by lloydjatkinson** [comment]
If we need to register file extension associations, or write other registry values, how would we do that with Velopack? Should we do that in one of the hooks?

**Reply by caesay** [reply]
We eventually wish to support file associations and custom url protocols in Velopack natively (https://github.com/velopack/velopack/issues/29) but until then, you can install and uninstall any registry entries needed via the after-install and before-uninstall hooks.

**Comment by kalvinet** [comment]
Hi. Do you plan to add GPG signing options for the AppImage files for linux?

**Reply by caesay** [reply]
We eventually plan to add package signing for all platforms but I can not provide any ETA
